### DISCUSSION TITLE
Tenant discriminator in M:N relationship

### DISCUSSION THREAD

**Discussion by evantrimboli** [mikro-orm/mikro-orm#2168] [Q&A]
How would I represent something like this in Mikro ORM? Specifically, the challenge is the final two foreign keys to ensure that the tenant/post and tenant/tag references are correct. Is `joinColumns` a path I should be looking at here?

```sql
CREATE TABLE "tenant" (
 "id" serial PRIMARY KEY
);

CREATE TABLE "post" (
 "id" serial PRIMARY KEY,
 "tenant_id" int NOT NULL REFERENCES "tenant"("id"),
 "label" text NOT NULL,
 UNIQUE ("id", "tenant_id")
);

CREATE TABLE "tag" (
 "id" serial PRIMARY KEY,
 "tenant_id" int NOT NULL REFERENCES "tenant"("id"),
 "label" text NOT NULL,
 UNIQUE ("id", "tenant_id")
);

CREATE TABLE "post_tag" (
 "post_id" int NOT NULL REFERENCES "post"("id"),
 "tag_id" int NOT NULL REFERENCES "tag"("id"),
 "tenant_id" int NOT NULL REFERENCES "tenant"("id"),
 PRIMARY KEY ("post_id", "tag_id", "tenant_id")
 FOREIGN KEY ("post_id", "tenant_id") REFERENCES "post"("id", "tenant_id"),
 FOREIGN KEY ("tag_id", "tenant_id") REFERENCES "tag"("id", "tenant_id")
);
```

For reference, the entities have this shape:

```ts
@Entity()
class Tenant {
  @PrimaryKey()
  id!: number;
}

@Entity()
class Tag {
  @PrimaryKey()
  id!: number;

  @ManyToOne()
  tenant!: Tenant;
}

@Entity()
class Post {
  @PrimaryKey()
  id!: number;

  @ManyToOne()
  tenant!: Tenant;
}

@Entity()
class PostTag {
  @ManyToOne({
    primary: true,
    onDelete: 'cascade'
  })
  post!: Post;

  @ManyToOne({
    primary: true,
    onDelete: 'cascade'
  })
  tag!: Tag;

  @ManyToOne()
  tenant!: Tenant;
}
```

**Comment by B4nan** [comment]
I don't think this is currently supported. FKs are generated based on the PKs, so if you want composite FK, it needs to be made by having composite PK that it points to, and you don't have that. On the other hand, you specify composite unique constrain on the columns - why don't you make them the PK instead?

It will be still probably not possible to do it exactly as you want, the tenant column would be there twice if done this way, once for each of the composite PKs (we could probably make that work internally quite easily, but only by changing the ORM code, I don't see any workaround). 

I guess we could allow configuring the FKs separately too, on the entity level.

**Comment by evantrimboli** [comment]
Thanks for the reply.

> why don't you make them the PK instead?

I probably should, in practice I'm using `uuid` for the types so the keys are guaranteed unique anyway.

> I guess we could allow configuring the FKs separately too, on the entity level.

This seems like a reasonable approach. I had thought about the ability to add on extra keys to existing relationships (ie. be able to specify the tenant as part of the `ManyToOne`), but that seems like it would be difficult to handle generically.

A type safe way to set the foreign keys would be cool. Just throwing out a starting point, it could look something like:

```ts
@Entity()
// Type safe
@ForeignKey<PostTag>([{
    field: 'post',
    reference: () => getReferenceToPostId()
}, {
    field: 'tenant',
    reference: () => getReferenceToPostTenantId()
}], [/* Repeat for tag  */])
export default class PostTag {
  @ManyToOne({
    primary: true,
    // Don't generate FK by default
    customForeignKey: true
  })
  post!: Post;

  @ManyToOne({
    primary: true,
    // Don't generate FK by default
    customForeignKey: true
  })
  tag!: Tag;

  @ManyToOne({
    primary: true,
    // Don't generate FK by default
    customForeignKey: true
  })
  tenant!: Tenant;

  [PrimaryKeyType]: [Post['id'], Tag['id']];
}
```

Not sure of the API to get a reference to the other fields via the `reference`. Would be interested to hear your thoughts.
